File: examples/SocialAppLayout-Inverted/Sample/LikesNode.m
//
//  LikesNode.m
//  Texture
//
//  Copyright (c) Facebook, Inc. and its affiliates.  All rights reserved.
//  Changes after 4/13/2017 are: Copyright (c) Pinterest, Inc.  All rights reserved.
//  Licensed under Apache 2.0: http://www.apache.org/licenses/LICENSE-2.0
//

#import "LikesNode.h"
#import "TextStyles.h"

@interface LikesNode ()
@property (nonatomic, strong) ASImageNode *iconNode;
@property (nonatomic, strong) ASTextNode *countNode;
@property (nonatomic, assign) NSInteger likesCount;
@property (nonatomic, assign) BOOL liked;
@end

@implementation LikesNode

- (instancetype)initWithLikesCount:(NSInteger)likesCount
{
    self = [super init];
    if (self) {
        _likesCount = likesCount;
        _liked = (_likesCount > 0) ? [LikesNode getYesOrNo] : NO;
        
        _iconNode = [[ASImageNode alloc] init];
        _iconNode.image = (_liked) ? [UIImage imageNamed:@"icon_liked.png"] : [UIImage imageNamed:@"icon_like.png"];
        [self addSubnode:_iconNode];
        
        _countNode = [[ASTextNode alloc] init];
        if (_likesCount > 0) {
            
            NSDictionary *attributes = _liked ? [TextStyles cellControlColoredStyle] : [TextStyles cellControlStyle];
            _countNode.attributedText = [[NSAttributedString alloc] initWithString:[NSString stringWithFormat:@"%ld", (long)_likesCount] attributes:attributes];
            
        }
        [self addSubnode:_countNode];
        
        // make it tappable easily
        self.hitTestSlop = UIEdgeInsetsMake(-10, -10, -10, -10);
    }
    
    return self;
    
}

+ (BOOL)getYesOrNo
{
    int tmp = (arc4random() % 30)+1;
    if (tmp % 5 == 0) {
        return YES;
    }
    return NO;
}

- (ASLayoutSpec *)layoutSpecThatFits:(ASSizeRange)constrainedSize
{
    ASStackLayoutSpec *mainStack =
    [ASStackLayoutSpec
     stackLayoutSpecWithDirection:ASStackLayoutDirectionHorizontal
     spacing:6.0
     justifyContent:ASStackLayoutJustifyContentStart
     alignItems:ASStackLayoutAlignItemsCenter
     children:@[_iconNode, _countNode]];

    mainStack.style.minWidth = ASDimensionMakeWithPoints(60.0);
    mainStack.style.maxHeight = ASDimensionMakeWithPoints(40.0);

    return mainStack;
}

@end

File: examples/SocialAppLayoutSwift/SocialAppLayoutSwift/LikesNode.swift
//
//  LikesNode.swift
//  SocialAppLayoutSwift
//
//  Created by Dicky Johan on 14/4/17.
//  Copyright © 2017 Dicky Johan. All rights reserved.
//

import Foundation
import AsyncDisplayKit

class LikesNode : ASControlNode {

    var iconNode:ASImageNode?
    var countNode:ASTextNode?
    var likesCount:Int?
    var liked:Bool?
    
    init(likesCount:Int) {
        super.init()
        self.likesCount = likesCount
        self.liked = (likesCount > 0) ? LikesNode.getYesOrNo() : false
        
        let iconNode = ASImageNode()
        iconNode.image = UIImage(named:"icon_liked.png")
        self.addSubnode(iconNode)
        self.iconNode = iconNode
        
        let countNode = ASTextNode()
        if (self.likesCount! > 0) {
            let attributes = self.liked! ? TextStyles.cellControlColoredStyle : TextStyles.cellControlStyle
            countNode.attributedText = NSAttributedString(string:String(format:"%ld",likesCount),attributes:attributes)
        }
        self.addSubnode(countNode)
        self.countNode = countNode
        
        // make it tappable easily
        self.hitTestSlop = UIEdgeInsetsMake(-10, -10, -10, -10)
        
    }
    
    static func getYesOrNo() -> Bool {
        let tmp = (arc4random() % 30)+1
        if (tmp % 5 == 0) {
            return true
        }
        return false
    }
    
    override func layoutSpecThatFits(_ constrainedSize: ASSizeRange) -> ASLayoutSpec {
        let mainStack = ASStackLayoutSpec(direction: .horizontal, spacing: 6.0, justifyContent: .start, alignItems: .center, children: [self.iconNode!, self.countNode!])
        
        mainStack.style.minWidth = ASDimensionMakeWithPoints(60.0)
        mainStack.style.maxHeight = ASDimensionMakeWithPoints(40.0)
        
        return mainStack
    }
    
}
